x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/ModelFieldType.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.core.ml.inference;
+
+import java.util.List;
+import java.util.Locale;
+
+public enum ModelFieldType {
+
+    SCALAR,
+    CATEGORICAL,
+    VECTOR,
+    TEXT;
+
+    public static ModelFieldType fromString(String value) {
+        return ModelFieldType.valueOf(value.toUpperCase(Locale.ROOT));
+    }
+
+    @Override
+    public String toString() {
+        return name().toLowerCase(Locale.ROOT);
+    }
+
+    public boolean supportsJavaObjectType(Object obj) {
+        if (obj == null) {
+            return false;
+        }
+        switch (this) {
+            case VECTOR:
+                return obj instanceof List || obj instanceof Object[];
+            case SCALAR:
+                if (obj instanceof Number) {
+                    return true;
+                }
+                return obj instanceof String && isDouble((String)obj);
+            case CATEGORICAL:
+                return obj instanceof String || obj instanceof Number;
```

Comment:
This definition of `categorical` is different to the one in data frame analytics: https://github.com/elastic/elasticsearch/blob/303b7e3e441e25f407609d34f5417b1201260299/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/Types.java#L26-L28

And not even the union of data frame analytics's `categorical` and `discrete_numerical`: https://github.com/elastic/elasticsearch/blob/303b7e3e441e25f407609d34f5417b1201260299/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/Types.java#L34-L37

Should we be aiming for consistency here?

Reply:
@droberts195 
Data frame analytics has the luxury of the mapped types. So, it can be more particular in its choices. Items coming into infer (right now) is effectively a mapped object of document `_source`. 

Something could be mapped as `keyword` but actually be a number in `_source`.

Example:
```
PUT numeric_categorical
{
  "mappings": {
    "properties": {
      "categorical": {
        "type": "keyword"
      }
    }
  }
}

POST numeric_categorical/_doc
{
  "categorical": 2
}
```